### DISCUSSION TITLE
Build failed under FreeBSD

### DISCUSSION THREAD

**Discussion by iron-udjin** [ghostty-org/ghostty#3832] [Q&A] [answered]
Is it possible to add support FreeBSD?

OS: FreeBSD 14-STABLE
zig: 0.13.0

```
$ zig build -Doptimize=ReleaseFast
warning: target=Target.Os.Tag.freebsd not supported
thread 101839 panic: unhandled error
/tmp/ghostty/pkg/sentry/build.zig:123:13: 0x12d7b87 in build (build)
            return error.UnsupportedTarget;
            ^
/usr/local/lib/zig/std/Build.zig:2117:24: 0x1232837 in runBuild__anon_19722 (build)
        .ErrorUnion => try build_zig.build(b),
                       ^
Unwind information for `libc.so.7:0x80163c860` was not available, trace may be incomplete

/usr/local/lib/zig/std/Build.zig:2097:40: 0x11bd043 in dependencyInner__anon_18700 (build)
        sub_builder.runBuild(bz) catch @panic("unhandled error");
                                       ^
/usr/local/lib/zig/std/Build.zig:1954:35: 0x111eebe in dependency__anon_11110 (build)
            return dependencyInner(b, name, pkg.build_root, if (@hasDecl(pkg, "build_zig")) pkg.build_zig else null, pkg_hash, pkg.deps, args);
                                  ^
/tmp/ghostty/build.zig:1242:36: 0x11150b7 in addDeps (build)
    const sentry_dep = b.dependency("sentry", .{
                                   ^
/tmp/ghostty/build.zig:1743:24: 0x11260ab in benchSteps (build)
        _ = try addDeps(b, c_exe, config: {
                       ^
/tmp/ghostty/build.zig:337:19: 0x112dc01 in build (build)
    try benchSteps(b, target, config, emit_bench);
                  ^
/usr/local/lib/zig/std/Build.zig:2117:43: 0x10eb2f4 in runBuild__anon_8128 (build)
        .ErrorUnion => try build_zig.build(b),
                                          ^
/usr/local/lib/zig/compiler/build_runner.zig:301:29: 0x10e5be8 in main (build)
        try builder.runBuild(root);
                            ^
/usr/local/lib/zig/std/start.zig:524:37: 0x10edbab in main (build)
            const result = root.main() catch |err| {
                                    ^
???:?:?: 0x801597ff3 in ??? (libc.so.7)
???:?:?: 0x10cb01f in ??? (???)
???:?:?: 0x8014c8007 in ??? (???)
error: the following build command crashed:
/tmp/ghostty/.zig-cache/o/4e089248bdda3c566012cffebc2efe05/build /usr/local/bin/zig /tmp/ghostty /tmp/ghostty/.zig-cache /home/iron/.cache/zig --seed 0xf7c8962b -Z40637ff6fdc049c8 -Doptimize=ReleaseFast
```

**Comment by 00-kat** [comment]
Some dependencies (most notably, Sentry, which is the one in the stack trace) don't support BSDs, sadly.

I'm not sure if `libxev` supports BSDs, but it does have kqueue for macOS so I doubt it would be particularly challenging to add support in the future if it is planned.

You could try FreeBSD's [Linuxulator](https://wiki.freebsd.org/Linuxulator) ([link to manual](https://docs.freebsd.org/en/books/handbook/linuxemu/)), but I don't know if it actually works (not sure if anyone's tried it yet).  Or maybe a Jail, but I doubt that would be a good solution.

**Reply by iron-udjin** [reply]
>You could try FreeBSD's Linuxulator

Linuxulator is not a good way to go. It's better to use native.

>I'm not sure if libxev supports BSDs

It doesn't build also:

```
$ zig build
install
└─ zig build-lib xev Debug native 1 errors
src/main.zig:46:13: error: no default backend for this target
            @compileError("no default backend for this target");
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/main.zig:6:28: note: called from here
const xev = Backend.default().Api();
            ~~~~~~~~~~~~~~~^~
referenced by:
    usingnamespace_0: src/main.zig:7:20
    remaining reference traces hidden; use '-freference-trace' to see all reference traces

Compile Log Output:
@as(Target.Os, .{ .tag = .freebsd, .version_range = .{ .semver = .{ .min = .{ ... }, .max = .{ ... } } } })
error: the following command failed with 1 compilation errors:
/usr/local/bin/zig build-lib -ODebug -Mroot=/tmp/libxev/src/c_api.zig -lc --cache-dir /tmp/libxev/.zig-cache --global-cache-dir /home/iron/.cache/zig --name xev -static --listen=-
install
└─ install xev
   └─ zig build-lib xev Debug native 1 errors
src/main.zig:46:13: error: no default backend for this target
            @compileError("no default backend for this target");
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/main.zig:6:28: note: called from here
const xev = Backend.default().Api();
            ~~~~~~~~~~~~~~~^~
referenced by:
    usingnamespace_0: src/main.zig:7:20
    remaining reference traces hidden; use '-freference-trace' to see all reference traces

Compile Log Output:
@as(Target.Os, .{ .tag = .freebsd, .version_range = .{ .semver = .{ .min = .{ ... }, .max = .{ ... } } } })
error: the following command failed with 1 compilation errors:
/usr/local/bin/zig build-lib -ODebug -Mroot=/tmp/libxev/src/c_api.zig --cache-dir /tmp/libxev/.zig-cache --global-cache-dir /home/iron/.cache/zig --name xev -dynamic --listen=-
Build Summary: 4/9 steps succeeded; 2 failed (disable with --summary none)
install transitive failure
├─ install xev transitive failure
│  └─ zig build-lib xev Debug native 1 errors
├─ zig build-lib xev Debug native (reused)
├─ install xev transitive failure
│  └─ zig build-lib xev Debug native 1 errors
└─ zig build-lib xev Debug native (reused)
error: the following build command failed with exit code 1:
/tmp/libxev/.zig-cache/o/d6885cb9139c84c9252f1d51fa8859c4/build /usr/local/bin/zig /tmp/libxev /tmp/libxev/.zig-cache /home/iron/.cache/zig --seed 0x83acefc1 -Za07196e996339bad
```

**Reply by rhodes-b** [reply]
~~*BSD is not a supported platform currently~~ edit: I see Kat got here first

**Reply by ghost** [reply]
so should we ask the maintainer of the dependency to add the bsd's?

**Reply by iron-udjin** [reply]
> so should we ask the maintainer of the dependency to add the bsd's?

Of course, if it's possible.

**Reply by ghost** [reply]
if noone else wants to make the feature request i would volenteer

**Reply by ghost** [reply]
https://github.com/getsentry/sentry-native/issues/1090

might this be the issue in sentry?

**Reply by jcollie** [reply]
There's a PR that should hopefully be merged soon that disables Sentry for anything but macOS. The more serious issue is getting BSD support in libxev as that's not an optional dependency. That's probably the best place to start for someone that's serious about getting Ghostty on BSD.

**Reply by ghost** [reply]
should someone who knows zig make the request?
i would but i dont think i know enough to talk about these issues

**Reply by ghost** [reply]
> There's a PR that should hopefully be merged soon that disables Sentry for anything but macOS. The more serious issue is getting BSD support in libxev as that's not an optional dependency. That's probably the best place to start for someone that's serious about getting Ghostty on BSD.

https://github.com/mitchellh/libxev/issues/67

**Reply by rbgarga** [reply]
https://github.com/getsentry/sentry-native/issues/1090 is now fixed

**Reply by mitchellh** [reply]
Thanks, if/when we work on FreeBSD in earnest then this won't be a blocker anymore at least. 😄

**Comment by mitchellh** [comment, marked as answer]
BSD isn't supported currently. I'd be happy to accept patches to fix this.

**Comment by taq** [comment]
Dude! I was trying since yesterday to compile the source code, installed Zig etc ... but I'm facing some trouble with my ISP and codeberg.org, where seems it is needed to download some Zig packages. Ok, it is weird stuff, I already contacted my ISP because after 6 hops the route is dead, but it was very useful to found this topic here so I won't try to compile it if it is not supported yet, as my main environment is Linux and I'm not so into FreeBSD to try to port it.

But let me ask: I'm not into Zig language and build system, but I saw that it tries to download a .gz file from CodeBerg, is there any checksum/security procedure to check if the download is secure/ok? I just found this download after it timed out on codeberg.org and showed my the error.

**Reply by mitchellh** [reply]
> But let me ask: I'm not into Zig language and build system, but I saw that it tries to download a .gz file from CodeBerg, is there any checksum/security procedure to check if the download is secure/ok? I just found this download after it timed out on codeberg.org and showed my the error.

Yes we have a sha256 checksum within our repository, directly in the `build.zig.zon` file.

**Reply by taq** [reply]
Thanks for the answer. This is a very important part of security.

**Comment by ghost** [comment]
i may be wrong but iirc libxev is needed for this and it just added support for freebsd: https://github.com/mitchellh/libxev/issues/67#event-18197848000

**Reply by 00-kat** [reply]
#7606.

**Reply by ghost** [reply]
thanks
